crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
         #[cfg(feature = "experimental-sliding-sync")]
         let mut sliding_sync_proxy: Option<Url> = None;
 
-        let homeserver = match homeserver_cfg {
-            HomeserverConfig::Url(url) => {
-                #[cfg(feature = "experimental-sliding-sync")]
-                {
-                    sliding_sync_proxy =
-                        self.sliding_sync_proxy.as_ref().map(|url| Url::parse(url)).transpose()?;
+        let (homeserver, well_known) = match homeserver_cfg {
+            HomeserverConfig::Url(url) => (url, None),
+            HomeserverConfig::ServerName { server: server_name, protocol } => {
+                let well_known = discover_homeserver(server_name, protocol, &http_client).await?;
+                (well_known.homeserver.base_url.clone(), Some(well_known))
+            }
+            HomeserverConfig::ServerNameOrUrl(server_name_or_url) => {
+                // Store the result to return at the end. If this doesn't get modified, then the
+                // supplied name is neither a server name, nor a valid URL.
+                let mut result: Result<
+                    (String, Option<discover_homeserver::Response>),
+                    ClientBuildError,
+                > = Err(ClientBuildError::InvalidServerName);
+
+                // Attempt discovery as a server name first.
+                let sanitize_result = sanitize_server_name(&server_name_or_url);
+                if let Ok(server_name) = sanitize_result.as_ref() {
+                    let protocol = if server_name_or_url.starts_with("http://") {
+                        UrlScheme::Http
+                    } else {
+                        UrlScheme::Https
+                    };
+
+                    result = match discover_homeserver(server_name.clone(), protocol, &http_client)
+                        .await
+                    {
+                        Ok(well_known) => {
+                            Ok((well_known.homeserver.base_url.clone(), Some(well_known)))
+                        }
+                        Err(e) => Err(e),
+                    }
                 }
 
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    allow_insecure_oidc = url.starts_with("http://");
+                // When discovery fails, or the input isn't a valid server name, fallback to
+                // trying a homeserver URL if supplied.
+                if let Ok(homeserver_url) = Url::parse(&server_name_or_url) {
```

Comment:
So the comment and code don't really match: here we should check *before* if `result.is_err()` in the current state, or with my proposal to use an option, if it's still none. I suppose that a valid homeserver name may not be a valid URL, but if they can, then this branch would override what we had done in the previous one, so let's make it clear by adding the `is_none()` check first.

Reply:
Oooops, yes good spot 🙏